execution-engine/contracts/swap/swap-proxy/src/client_api/mod.rs

```diff
@@ -0,0 +1,237 @@
+mod error;
+
+use alloc::{string::String, vec::Vec};
+
+use contract::{contract_api::runtime, unwrap_or_revert::UnwrapOrRevert};
+use types::{account::PublicKey, ApiError, ContractRef, Key, U512};
+
+use error::Error;
+
+pub mod method_names {
+    pub mod proxy {
+        use super::swap;
+
+        pub const NAME_SWAP_HASH: &str = "swap_hash";
+        pub const METHOD_SET_SWAP_HASH: &str = swap::METHOD_SET_SWAP_HASH;
+        pub const METHOD_INSERT_KYC_ALLOWANCE_CAP: &str = swap::METHOD_INSERT_KYC_ALLOWANCE_CAP;
+        pub const METHOD_INSERT_SNAPSHOT_RECORD: &str = swap::METHOD_INSERT_SNAPSHOT_RECORD;
+        pub const METHOD_INSERT_KYC_DATA: &str = swap::METHOD_INSERT_KYC_DATA;
+        pub const METHOD_UPDATE_KYC_LEVEL: &str = swap::METHOD_UPDATE_KYC_LEVEL;
+        pub const METHOD_UPDATE_STATUS_SWAPABLE_TOKEN_SENT: &str =
+            swap::METHOD_UPDATE_STATUS_SWAPABLE_TOKEN_SENT;
+        pub const METHOD_UPDATE_KYC_STEP: &str = swap::METHOD_UPDATE_KYC_STEP;
+        pub const METHOD_GET_TOKEN: &str = swap::METHOD_GET_TOKEN;
+    }
+    pub mod swap {
+        pub const METHOD_SET_SWAP_HASH: &str = "set_swap_hash";
+        pub const METHOD_INSERT_KYC_ALLOWANCE_CAP: &str = "insert_kyc_allowance_cap";
+        pub const METHOD_INSERT_SNAPSHOT_RECORD: &str = "insert_snapshot_record";
+        pub const METHOD_INSERT_KYC_DATA: &str = "insert_kyc_data";
+        pub const METHOD_UPDATE_KYC_LEVEL: &str = "update_kyc_level";
+        pub const METHOD_UPDATE_STATUS_SWAPABLE_TOKEN_SENT: &str =
+            "update_status_swapable_token_sent";
+        pub const METHOD_UPDATE_KYC_STEP: &str = "update_kyc_step";
+        pub const METHOD_GET_TOKEN: &str = "get_token";
+    }
+}
+
+pub enum Api {
+    SetSwapHash(Key),
+    InsertKYCAllowanceCap(U512),
+    InsertSnapshotRecord(String, U512),
+    InsertKYCData(PublicKey),
+    UpdateKYCLevel(PublicKey, U512),
+    UpdateStatusSwapableTokenSent(PublicKey, U512),
+    UpdateKYCStep(PublicKey, U512),
+    GetToken(Key, Vec<String>, Vec<String>, Vec<String>, Vec<String>),
+}
+
+fn get_contract_ref() -> ContractRef {
+    let contract_hash = runtime::get_key(method_names::proxy::NAME_SWAP_HASH)
+        .unwrap_or_revert_with(ApiError::GetKey);
+    contract_hash.to_contract_ref().unwrap_or_revert()
```

Comment:
These can be simplified like below,
```Rust
fn get_contract_ref() -> ContractRef {
    runtime::get_key(method_names::proxy::NAME_SWAP_HASH)
        .unwrap_or_revert_with(ApiError::GetKey);
        .to_contract_ref()
        .unwrap_or_revert()
}
```

Also, this proxy requires that the caller has the NAME_SWAP_HASH in their named key.
This requires a user to run the contract storing NAME_SWAP_HASH in their named_key.
Why don't we allow a user just to call by hash without storing it under their named_key?

More than that, how about integrating the logic, proxy, and install into a single contract?

Reply:
> Why don't we allow a user just to call by hash without storing it under their named_key?
Do you mean to get hash as a parameter in contract call?

Reply:
> More than that, how about integrating the logic, proxy, and install into a single contract?

Try to do that but assign to lower priority.